adr/ADR-32.md

```diff
@@ -0,0 +1,179 @@
+
+The introduction of JetStream brought an API which provides the ability to manage and interact with streams and consumers. This API was intended to be _internal_ given the design of the subjects. Client libraries have implemented constructs over top the raw API calls. However, the these higher level constructs never manifested as user permissions.
+
+The concern is that developers need to model permissions based on the internal JS subjects which is both difficult, but also introduces coupling between the developer's application-level need and the API design.
+
+## [Context | References | Prior Work]
+
+[Permissions][1] are defined for publish or subscribe and list a set of subjects (including wildcards). As noted above, `allow_responses` is an abstraction to handle the "reply to any inbox" situation.
+
+With the introduction of JetStream, APIs such as `add-stream`, `delete-stream`, etc. were introduced. These reduce down to (mostly) a set of request-reply interactions. For example, to create a stream, a request must be made to `$JS.API.STREAM.CREATE.<name>` where `<name>` is the named of the stream. To get consumer info a request to `$JS.API.CONSUMER.INFO.<stream>.<name>` must be allowed.
+
+The question is, why does a developer need to know about these verbose subjects when defining permissions?
+
+[1]: https://docs.nats.io/running-a-nats-service/configuration/securing_nats/authorization#permissions-configuration-map
+
+## Design
+
+Fundamentally, this design is a light abstraction on top of the current way permissions are defined in configuration or JWTs. These new constructs can be converted to a standard permissions map internally.
+
+There are two potential layers to defining user permissions. First, a logical name must be defined for the underlying API subject. This mapping is shown in the table below. In addition to the name, the _context_ (publish or subscribe) can also be inferred. For example, all of the JS APIs use request-reply initiated by the client. So `js-create-stream` is implicitly a publish-based permission. For subjects that are *parameterized* based on stream names, consumer names, or subjects, the `(...)` syntax is used to support specifying these parameters.
+
+Category | Name    | Subject | Pub/Sub | Notes
+---------|---------| ------- | ------- | -----
+General  | `pub(subject)` | `{subject}` | pub | Potential alternative to a bare subject string.
+\* | `sub(subject)` | `{subject}` | sub | Same as above.
+\* | `inbox` | `_INBOX.>` | sub
+\* | `inbox(id)` | `_INBOX_{id}.>` | sub | *Convention for custom inboxes, so be used with `--inbox-prefix` for private access.*
```

Comment:
need to be able to express the permissions for ACKing responses from the consumer, maybe there could be a macro for js-consumer to be able to subscribe and ack

Reply:
[These](https://github.com/nats-io/nats-architecture-and-design/pull/145/files#diff-c13b1651f7f36a6775ce53f919367c98427da78d1b669f27a90eae0423beab2aR92-R95) correct?

Reply:
ah yes those...